projects/core/src/routing/configurable-routes/configurable-routes.service.ts

```diff
@@ -49,75 +47,102 @@ export class ConfigurableRoutesService {
         return {
           ...route,
-          matcher: this.urlMatcherFactory.getFalsyUrlMatcher(),
+          matcher: this.urlMatcherService.getFalsy(),
         };
       } else if (matchers) {
         delete route.path;
-        return { ...route, matcher: this.resolveUrlMatchers(matchers) };
+        return { ...route, matcher: this.resolveUrlMatchers(route, matchers) };
       } else if (paths.length === 1) {
         delete route.matcher;
         return { ...route, path: paths[0] };
       } else {
         delete route.path;
         return {
           ...route,
-          matcher: this.urlMatcherFactory.getMultiplePathsUrlMatcher(paths),
+          matcher: this.urlMatcherService.fromPaths(paths),
         };
       }
     }
-    return route; // if route doesn't have a name, just pass the original route
+    return route; // if route doesn't have a Spartacus' name, just pass the original route
   }
 
-  private resolveUrlMatchers(matchers: RouteConfig['matchers']): UrlMatcher {
-    const createdMatchers: UrlMatcher[] = matchers.map(matcher => {
-      return typeof matcher === 'function'
-        ? matcher
-        : this.urlMatcherFactory.createUrlMatcher(matcher);
+  private resolveUrlMatchers(
+    route: Route,
+    matchersOrFactories: RouteConfig['matchers']
+  ): UrlMatcher {
+    const matchers: UrlMatcher[] = matchersOrFactories.map(matcherOrFactory => {
+      return typeof matcherOrFactory === 'function'
+        ? matcherOrFactory // matcher
+        : this.resolveUrlMatcherFactory(route, matcherOrFactory); // factory injection token
     });
-    const result = this.urlMatcherFactory.combineUrlMatchers(createdMatchers);
-    return result;
+    return this.urlMatcherService.combine(matchers);
+  }
+
+  private resolveUrlMatcherFactory(
+    route: Route,
+    factoryToken: InjectionToken<UrlMatcherFactory>
+  ): UrlMatcher {
+    const factory = this.injector.get(factoryToken);
+    return factory(route);
   }
 
   private getRouteName(route: Route): string {
     return route.data && route.data.cxRoute;
   }
 
-  private getConfiguredPaths(
+  private validateRouteConfig(
     routeConfig: RouteConfig,
     routeName: string,
     route: Route
-  ): string[] {
-    if (routeConfig === undefined) {
-      this.warn(
-        `Could not configure the named route '${routeName}'`,
-        route,
-        `due to undefined key '${routeName}' in the routes config`
-      );
-      return [];
-    }
-    if (routeConfig && routeConfig.paths === undefined) {
-      this.warn(
-        `Could not configure the named route '${routeName}'`,
-        route,
-        `due to undefined 'paths' for the named route '${routeName}' in the routes config`
-      );
-      return [];
-    }
+  ) {
+    if (isDevMode()) {
+      // route config being undefined is a misconfiguration
+      if (routeConfig === undefined) {
+        this.warn(
+          `Could not configure the named route '${routeName}'`,
+          route,
```

Comment:
I'd interpolate the `route` as well to make the code better readable.

Reply:
Unfortunately, then it will be printed as `Could not configure the named route 'product' [object Object] ...`. Making it Separate argument of `console.warn(, , , )` allows for inspecting the object.